File: ecocircuit/lib/models/order.dart
import 'product.dart';

class CartItem {
  final Product product;
  int quantity;

  CartItem({
    required this.product,
    this.quantity = 1,
  });

  double get totalPrice => product.price * quantity;
}

class DeliveryAddress {
  final String fullName;
  final String phone;
  final String addressLine;
  final String city;
  final String pincode;

  const DeliveryAddress({
    required this.fullName,
    required this.phone,
    required this.addressLine,
    required this.city,
    required this.pincode,
  });
}

class UserOrder {
  final String id;
  final List<CartItem> items;
  final DateTime orderDate;
  final double subtotal;
  final double deliveryCharge;
  final double totalAmount;
  final String status; // 'Order Placed', 'Confirmed', 'Shipped', 'Delivered', 'Cancelled'
  final DeliveryAddress deliveryAddress;
  final String paymentMethod;

  const UserOrder({
    required this.id,
    required this.items,
    required this.orderDate,
    required this.subtotal,
    this.deliveryCharge = 49.0,
    required this.totalAmount,
    this.status = 'Order Placed',
    required this.deliveryAddress,
    this.paymentMethod = 'Cash on Delivery (Eco-Verified)',
  });
}

File: backend/src/models/Order.js
import mongoose from 'mongoose';

const orderSchema = new mongoose.Schema(
  {
    orderId: {
      type: String,
      unique: true,
    },
    buyerId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
    },
    sellerId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'User',
    },
    listingId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Listing',
      required: true,
    },
    type: {
      type: String,
      enum: ['direct_purchase', 'offer_negotiation', 'part_purchase'],
      default: 'direct_purchase',
    },
    amount: {
      type: Number,
      required: true,
      min: 0,
    },
    paymentMethod: {
      type: String,
      enum: ['upi', 'card', 'cod', 'wallet_points'],
      default: 'upi',
    },
    paymentStatus: {
      type: String,
      enum: ['pending', 'completed', 'failed', 'refunded'],
      default: 'pending',
    },
    deliveryAddress: {
      name: String,
      phone: String,
      addressLine: String,
      city: String,
      pincode: String,
    },
    status: {
      type: String,
      enum: ['order_placed', 'confirmed', 'packed', 'shipped', 'out_for_delivery', 'delivered', 'cancelled'],
      default: 'order_placed',
    },
    trackingNumber: String,
  },
  {
    timestamps: true,
  }
);

orderSchema.pre('save', function (next) {
  if (!this.orderId) {
    const randomSuffix = Math.floor(100000 + Math.random() * 900000);
    this.orderId = `ORD-2026-${randomSuffix}`;
  }
  next();
});

export const Order = mongoose.model('Order', orderSchema);
